documentation/Action-plugin-specification.md

```diff
@@ -0,0 +1,133 @@
+- `function_name` -- the name of the function exported in the source code file. If this name is changed in the source file, it _must_ be updated here, or else the app won't find the function.
+- `required_parameters` -- an array of strings that define the names of the fields that must be supplied to the plugin at runtime. Templates store the values (or dynamic expressions) that map to these field names.
+
+## How plugins are loaded.
+
+At startup, the server app scans the `plugins` folder. In each plugin's subfolder, it looks for a `plugin.json` file, from which it reads the metadata. It compares this information with the list of plugins registered in `action_plugin` and updates the database accordingly:
+
+- Adds a new record for newly-found plugins
+- Updates plugin details that have changed
+- Removes records for missing plugins
+
+The `action_plugin` table is the primary record of what actions are available to the system, with the following fields stored:
+
+- `code`
+- `name`
+- `description`
+- `path` (determined by the folder the plugin was found in, plus the specified filename)
+- `function_name`
+- `required_parameters`
+
+<a name="development"></a>
+
+## Development and build process
+
+Plugins are intended to be stand-alone units, in that a user should be able to simply copy a plugin folder to another system and have it work straight away. (Eventually, we envisage that plugins could be imported directly via the front-end UI.) To that end, each plugin is developed as its own package, with its own `package.json` and `node_modules` folder, as well as Typescript and testing configurations.
+
+When developing a plugin, you can run the main app with `yarn dev` as normal, and changes you make to the plugin `.ts` file will be reflected immediately in the dev environment.
+
+However, when development work is complete, the plugin should be built independently by running `yarn build` _in the plugin's root folder_, which will compile the typescript code into `.js` file(s). (There is also a shortcut `yarn build_plugins` to build all plugins from the project root.) When the main project is built (`yarn build`), the plugins are not re-compiled, but simply copied directly to the build folder, so its expected they will each have their own compiled `.js` file already in place.
+
+## Needs consideration
+
+- Should the `required_parameters` specify an expected type so that the Template Builder UI knows what kind of options to present?
+
+- Should there be (optional) **default values** for the `required_parameters`? I'm thinking, for things like "Send notification", the user's email address will be required, but it's always going to be the same query (`{operator: "objectProperties", children: [{value: {object: "user", property: "email"}}]}`) so it seems like it should there by default so the Admin doesn't have to specify an obvious value in the Template Builder.
```

Comment:
Yes. Indeed. I'm just trying to figure how we can check if the default values are correct. When the person developing the plugin we would give some sort of error as: `"this objectProperty is not defined in this context"` But I agree that those very basic objects should be defined wiht a default value. And in case that is changed it would just be passed in the `parameters` by something defined in the application template queries, right?

Reply:
I think plugin developers (i.e. us! for now) should be expected to handle this correctly. But yes, an invalid query should return something like what you suggested.

And yeah, the "default" value would be the "placeholder" value shown in the Template Builder UI when selecting this Action.